### PR TITLE
Add root type constraint between KeyPath expressions and applications

### PR DESCRIPTION
**Pull request by mdiep**
This key path application was failing to resolve the key path:

```swift
_ = ""[keyPath: \.count]
```

This adds a constraint between the type the KeyPath application is subscripting and the base type of the KeyPath expression inside of it.

Resolves [SR-7380](https://bugs.swift.org/browse/SR-7380).

### CODE REVIEW

File: include/swift/AST/Expr.h [hunk 1]
```diff
@@ -2300,7 +2300,7 @@ class SubscriptExpr final : public LookupExpr,
     return Bits.SubscriptExpr.HasArgLabelLocs;
   }
 
-  /// Whether this call with written with a trailing closure.
+  /// Whether this call was written with a trailing closure.
```
Comment (slavapestov) [inline include/swift/AST/Expr.h]: This should be a separate patch - if the other fix gets reverted for whatever reason your typo fix will live on :)

**Comment by slavapestov** [comment]
@jckarter Want to take a look?

File: lib/Sema/CSSimplify.cpp [hunk 2]
```diff
@@ -4960,6 +4960,28 @@ ConstraintSystem::addKeyPathApplicationConstraint(Type keypath,
...
+  if (auto subscript = dyn_cast_or_null<SubscriptExpr>(anchor)) {
+    if (path.size() == 1 &&
+        path[0].getKind() == ConstraintLocator::SubscriptMember) {
+      if (auto indexTuple = dyn_cast<TupleExpr>(subscript->getIndex())) {
+        if (indexTuple->getNumElements() == 1) {
+          if (auto keyPathExpr = dyn_cast<KeyPathExpr>(indexTuple->getElement(0))) {
+            for (auto constraint : InactiveConstraints) {
+              if (constraint.getKind() == ConstraintKind::KeyPath &&
+                  constraint.getLocator()->getAnchor() == keyPathExpr) {
+                addConstraint(ConstraintKind::Subtype, root, constraint.getSecondType(), locator);
```
Comment (CodaFi) [inline lib/Sema/CSSimplify.cpp]: This typing rule seems to be duplicating the existing rule in CSGen (after drilling).  Try looking at what `visitUnresolvedMemberExpr` is doing.

File: include/swift/AST/Expr.h (hunk 1 above)
Comment (mdiep) [reply include/swift/AST/Expr.h]: Submitted as https://github.com/apple/swift/pull/17098.

File: lib/Sema/CSSimplify.cpp (hunk 2 above)
Comment (mdiep) [reply lib/Sema/CSSimplify.cpp]: I'm afraid I don't understand which existing rule you're referring to or what specifically in `visitUnresolvedMemberExpr` you're pointing me to. 😞 I'll need some more help to understand. Sorry!

Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: I think we should try our best to do this in CSGen because we are trying to avoid dealing with expressions in the solver at all costs...

Comment (mdiep) [reply lib/Sema/CSSimplify.cpp]: In CSGen you don't know that this is a KeyPath application—it's just a subscript. Because this is inside the subscript overload, I don't think it can be in CSGen. Or maybe I missing something?

But it should be possible to move it out a level to `ConstraintSystem::resolveOverload` if that's preferable.

Comment (CodaFi) [reply lib/Sema/CSSimplify.cpp]: > In CSGen you don't know that this is a KeyPath application

I think it can be done.  A naive way would be to add an extra "faux root" expression node where you can stash the type variable generated for the `root` expression in the KeyPath.  Then in the post-walk, you can recreate this logic (minus the locator digging) and directly bind the subscript's base to the faux root.  For example (don't actually do this, please clean it up):

```cxx
// ConstraintWalker::walkToExprPost
      if (auto subscript = dyn_cast<SubscriptExpr>(expr)) {
        if (auto indexTuple = dyn_cast<TupleExpr>(subscript->getIndex())) {
          if (indexTuple->getNumElements() == 1) {
            if (auto keyPathExpr = dyn_cast<KeyPathExpr>(indexTuple->getElement(0))) {
              auto type = CG.visit(expr);
              auto &CS = CG.getConstraintSystem();
              auto simplifiedType = CS.simplifyType(type);

              CS.setType(expr, simplifiedType);

              CS.addConstraint(ConstraintKind::Subtype,
                               CS.getType(subscript->getBase()),
                               CS.getType(keyPathExpr->getFauxRoot()),
                               CS.getConstraintLocator(expr));

              return expr;
            }
          }
        }
      }

// ...

// ConstraintWalker::visitKeyPathExpr
case KeyPathExpr::Component::Kind::UnresolvedProperty:
          E->setFauxRoot(/**/);
          CS.setType(E->getFauxRoot(), root);
          LLVM_FALLTHROUGH;
```

**Review by rudkx — COMMENTED** [review]
I still need to take a closure look at the code changes.

File: lib/Sema/CSSimplify.cpp (hunk 2 above)
Comment (mdiep) [reply lib/Sema/CSSimplify.cpp]: That assumes that a `[keyPath: <key path>]` expression _is_ a KeyPath application. I was assuming that shouldn't be hardcoded, but it actually is today. i.e., this fails to compile:

```swift
extension String {
	subscript(keyPath: KeyPath<Int, String>) -> String {
		return "🙈"
	}
}

_ = ""[keyPath: \Int.description]
```

Since that's apparently not valid, it should be fine to add this constraint outside of the KeyPath application constraint. (Although that kinda seems like a bug to me. Why shouldn't you be able to overload a subscript that takes a KeyPath?)

Given that…

1. Is there any reason it's preferable to use `walkToExprPost` over `visitSubscriptExpr`? As long as we're going to assume that `[keyPath: <key path>]` is a key path application, we might as well remove the overload and do this directly in `visitSubscriptExpr`?

2. Is it still preferable to create another expression instead of using the locator? Neither seem ideal to me. 😕 That would mean adding a new type like `KeyPathRootExpr`, right?

File: test/expr/unary/keypath/keypath.swift [hunk 3]
```diff
@@ -499,6 +499,10 @@ func sr6744() {
     _ = get(for: \.value)
 }
 
+func sr7380() {
+  _ = ""[keyPath: \.count]
+}
+
```
[the remark this answers is not in the document]
Comment (mdiep) [reply test/expr/unary/keypath/keypath.swift]: Great suggestion. The `var` case wasn't working; I've fixed that up.

**Comment by jckarter** [comment]
Looks good, thanks! When I tried this originally, I had issues getting the type system to keep working in the face of `PartialKeyPath` and maintain the correct mutability relationships for `Writable` and `ReferenceWritable` key paths. I saw Mark had you add some for writable key paths, but having some for class instance property key paths would be nice too, and it'd be a good idea to also add tests to ensure that both literal key paths and variables work, as well as PartialKeyPaths.

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

File: lib/Sema/CSSimplify.cpp [hunk 4]
```diff
@@ -4960,6 +4960,33 @@ ConstraintSystem::addKeyPathApplicationConstraint(Type keypath,
...
+  if (auto subscript = dyn_cast_or_null<SubscriptExpr>(anchor)) {
+    if (path.size() == 1 &&
+        path[0].getKind() == ConstraintLocator::SubscriptMember) {
+      if (auto indexTuple = dyn_cast<TupleExpr>(subscript->getIndex())) {
+        if (indexTuple->getNumElements() == 1) {
+          if (auto keyPathExpr = dyn_cast<KeyPathExpr>(indexTuple->getElement(0))) {
+            for (auto constraint : InactiveConstraints) {
+              if (constraint.getKind() == ConstraintKind::KeyPath &&
+                  constraint.getLocator()->getAnchor() == keyPathExpr) {
+                auto keyPathRootTy = constraint.getSecondType();
```
Comment (CodaFi) [inline lib/Sema/CSSimplify.cpp]: Collapse this

```cxx
addConstraint(ConstraintKind::Subtype, lvt->getWithoutSpecifierType(), keyPathRootTy, locator);
```

**Comment by mdiep** [comment]
> but having some for class instance property key paths would be nice too,

I added a couple tests for this. This error was even worse before:

```
7380.swift:23:5: error: type 'A' has no subscript members
_ = A()[keyPath: \.a]
    ^~~
```

> it'd be a good idea to also add tests to ensure that both literal key paths and variables work, as well as PartialKeyPaths

I added some tests for `\Type.property` key paths, but variables seem well tested already. (And since those are type-checked separately, they're pretty orthogonal to this bug and fix.)

File: lib/Sema/CSSimplify.cpp (hunk 2 above)
Comment (CodaFi) [reply lib/Sema/CSSimplify.cpp]: > Is there any reason it's preferable to use walkToExprPost over visitSubscriptExpr

I use it out of convenience.  Anywhere in the walk that has access to the subscript and its children after they have had constraints generated will do.

> As long as we're going to assume that [keyPath: <key path>] is a key path application, we might as well remove the overload and do this directly in visitSubscriptExpr

No such luck in the general case

```swift
extension String {
  subscript<T>(keypath a: T) -> String {
    return "🙈"
  }
}
_ = ""[keypath: 5]
```

> Is it still preferable to create another expression instead of using the locator

Digging in the locators is a hack, adding a fake AST node is a hack.  Neither is desirable.  You shouldn't need a new type for this, the whole goal of the example I gave was that you need to stash the type variable generated for root somewhere that's accessible to the rest of CSGen.

**Comment by jckarter** [comment]
How bad was the diagnostic before? @xedin and @rudkx might be able to advise on how to win back some diagnostic quality.

**Comment by mdiep** [comment]
Oh, sorry, I meant that `_ = A()[keyPath: \.a]` had that nonsense error before but now compiles with this fix.

File: lib/Sema/CSSimplify.cpp (hunk 2 above)
Comment (xedin) [reply lib/Sema/CSSimplify.cpp]: The suggestion @CodaFi made it along the lines I was thinking as well, but I'm not sure if we really need to create a faux root through, because even if we still let root allocate new type variable but then connect it to something which can inferred that would still be good enough. Another idea, but I'm not sure how feasible that would be, is to actually make it so `KeyPathExpr`, when created, knows what expression it's attached to, that would be very straightforward to get its type from the constraint system cache and relate it to root in `visitKeyPathExpr`.

File: lib/Sema/CSSimplify.cpp [hunk 5]
```diff
@@ -4960,6 +4960,29 @@ ConstraintSystem::addKeyPathApplicationConstraint(Type keypath,
...
+  // connects the subscript's root type to the root type of the KeyPath.
+  SmallVector<LocatorPathElt, 4> path;
+  Expr *anchor = locator.getLocatorParts(path);
+  if (auto subscript = dyn_cast_or_null<SubscriptExpr>(anchor)) {
+    if (path.size() == 1 &&
+        path[0].getKind() == ConstraintLocator::SubscriptMember) {
+      if (auto indexTuple = dyn_cast<TupleExpr>(subscript->getIndex())) {
+        if (indexTuple->getNumElements() == 1) {
+          if (auto keyPathExpr = dyn_cast<KeyPathExpr>(indexTuple->getElement(0))) {
+            for (auto constraint : InactiveConstraints) {
```
Comment (rudkx) [inline lib/Sema/CSSimplify.cpp]: Rather than iterating over all constraints here, and checking the anchor expression, you should be able to do something like:
```
  auto *typevar = cs.getType(keyPathExpr)->getAs<TypeVariableType>();
  if (typevar) {
    CG.gatherConstraints(typevar, constraints, ConstraintGraph::GatheringKind::EquivalenceClass);
    // iterate over only those looking for a ConstraintKind::KeyPath where the constraint.getFirstType()->isEqual(typevar)
  }
```

File: lib/Sema/CSSimplify.cpp [hunk 6]
```diff
@@ -4960,6 +4960,29 @@ ConstraintSystem::addKeyPathApplicationConstraint(Type keypath,
...
+  // If this is a subscript with a KeyPath expression, add a constraint that
+  // connects the subscript's root type to the root type of the KeyPath.
+  SmallVector<LocatorPathElt, 4> path;
+  Expr *anchor = locator.getLocatorParts(path);
+  if (auto subscript = dyn_cast_or_null<SubscriptExpr>(anchor)) {
+    if (path.size() == 1 &&
+        path[0].getKind() == ConstraintLocator::SubscriptMember) {
+      if (auto indexTuple = dyn_cast<TupleExpr>(subscript->getIndex())) {
+        if (indexTuple->getNumElements() == 1) {
+          if (auto keyPathExpr = dyn_cast<KeyPathExpr>(indexTuple->getElement(0))) {
```
Comment (rudkx) [inline lib/Sema/CSSimplify.cpp]: I suspect some of the above logic could be turned into `cast<>()` and `assert`s based on our expectations of what we know when we decide an overload is a key path application.

For whatever remains, it might be nice to remove some of the depth of control flow via the advice in https://www.llvm.org/docs/CodingStandards.html#use-early-exits-and-continue-to-simplify-code

You could for example make this a closure with early returns, a separate function, or a:
```
  do {
    // use break in places where we bail out
  } while (0);

**Review by rudkx — CHANGES_REQUESTED** [review]
Looks good. A couple small notes of things you should take a look at.

It would be great if we didn't have to iterate over constraints to do this, but offhand I cannot think of how we would do that.

**Comment by jckarter** [comment]
@swift-ci Please smoke test

**Comment by jckarter** [comment]
@swift-ci Please test source compatibility

**Comment by mdiep** [comment]
How does that look, @rudkx?

**Review by rudkx — APPROVED** [review]
LGTM!

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by rudkx** [comment]
@swift-ci Please test source compatibility

**Comment by mdiep** [comment]
I had to switch the check for a tuple back to a `dyn_cast`.

**Comment by rudkx** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
Worth cherry-picking into 4.2?

**Comment by jckarter** [comment]
Sounds good to me!
